In [9]:
import numpy as np 
import math 

# Ý tưởng phương pháp 

## Điều kiện HT của phương pháp 

## Tốc độ HT của phương pháp 

## Đánh giá sai số: 
Có hai cách đánh giá sai số: 
1. Công thức tiên nghiệm (đánh giá sai số trước khi tính giá trị xấp xỉ) 
2. Công thức hậu nghiệm (đánh giá sai số sau khi đã tính nghiệm xấp xỉ) 

- Câu hỏi đặt ra: Phương pháp nào cho tốc độ hội tụ đến nghiệm chính xác nhanh hơn?\
Hiện tại mình đang nghiêng về phương pháp tiên nghiệm, do nó đánh giá sai số dựa vào nghiệm chính xác của hàm. 

# Một số lưu ý về phương pháp chia đôi: 
- Nên sử dụng một biến đếm số vòng lặp là bao nhiêu. 
- Có một phương pháp không phải tính f(c) rồi nhân với f(a) hoặc f(b) để xác định dấu (hình như có liên quan đến định lý giá trị trung gian). Cố thử suy nghĩ về nó xem, nếu được thì không phải thực hiện phép tính f(c) và phép nhân xác định dấu ngay sau đó. 
- Sử dụng hai công thức đánh giá sai số là công thức tiên nghiệm và công thức hậu nghiệm. 

# Ý tưởng thuật toán: 

# Thuật toán: 

- ### Bước 1: 
- ### Bước 2: 
- ### Bước 3: 
- ### Bước 4: 

# Cài đặt thuật toán: 

## Trường hợp 1: Công thức hậu nghiệm (cho sai số, thực hiện xong phép lặp rồi đánh giá sai số) 

## Định nghĩa hàm f(x):

In [27]:
def f1(x): 
    return x**3 + 4*x**2 - 10

## Định nghĩa hàm co đoạn [a, b] còn một nửa: 

In [28]:
def Contraction1(a, b): 
    p = (a + b)/2
    
    if f(p) != 0: 
        if f(a)*f(p) > 0: # a, p cùng dấu 
            a = p 
        else: 
            b = p 
        return (a, b, p) 
    else: 
        printf("Nghiệm của phương trình đã cho là: %s." %(p))
        return (a, b, p) 

## Định nghĩa hàm thực thi chính: 

In [29]:
def BisectionMethod1(a, b, eps): 
    # định nghĩa biến đếm số vòng lặp. 
    iter = 0 
    
    (a, b, p1) = Contraction1(a, b)
    (a, b, p2) = Contraction1(a, b)
    
    while np.abs(p2 - p1) >= eps: 
        p1 = p2 
        (a, b, p2) = Contraction1(a, b)
        iter += 1
        
        # kiểm tra TH ngoại lệ: f(p) = 0 hay không (rất hiếm xảy ra với nghiệm xấp xỉ)
        if f(p2) != 0: 
            continue 
        else: 
            break
        
    return (p2, iter)

## Thực hiện chương trình: 

In [30]:
a = float(input("Nhập vào cận dưới: "))
b = float(input("Nhập vào cận trên: "))
eps = float(input("Nhập vào sai số mong muốn: "))

(p2, iter) = BisectionMethod1(a, b, eps)

print('Nghiệm x = %.12f, giá trị f(x) = %.12f, thu được sau %d vòng lặp.' %(p2, f1(p2), iter))

Nhập vào cận dưới: 1
Nhập vào cận trên: 2
Nhập vào sai số mong muốn: 1e-10
Nghiệm x = 1.365230013442, giá trị f(x) = 0.000000000457, thu được sau 32 vòng lặp.


- **Nhận xét: Nghiệm giải được trên mạng là: 1.365230013... => có thể nhận thấy độ chính xác mà ta tính được là khá lớn, đúng đến 9 chữ số sau dấu phẩy. (đúng hơn cả máy tính casio)**

## Trường hợp 2: Công thức tiên nghiệm. 

In [12]:
def f2(x): 
    return x**3 + 4*x**2 - 10

In [13]:
def Contraction2(a, b): 
    p = (a + b)/2
    
    if f(p) != 0: 
        if f(a)*f(p) > 0: # a, p cùng dấu 
            a = p 
        else: 
            b = p 
        return (a, b, p) 
    
    return (a, b, p) 

In [33]:
def BisectionMethod2(a, b, eps): 
    # định nghĩa biến đếm số vòng lặp.
    
    # giải bất đẳng thức để xác định số lần chia.
    iter = round(np.log2((b - a)/eps) + 0.5) 
    
    for i in range(iter): 
        (a, b, p) = Contraction2(a, b)
        
        # kiểm tra TH ngoại lệ: f(p) = 0 hay không (rất hiếm xảy ra với nghiệm xấp xỉ)
        if f(p) != 0: 
            continue 
        else: 
            break
        
    return (p, iter)

### Thực hiện chương trình: 

In [34]:
(p, iter) = BisectionMethod2(1, 2, 1e-10)
print('Nghiệm x = %.12f, giá trị f(x) = %.12f, thu được sau %d vòng lặp.' %(p, f2(p), iter))

Nghiệm x = 1.365230013442, giá trị f(x) = 0.000000000457, thu được sau 34 vòng lặp.


# Thuyết trình 

- Làm một cái bảng ghi kết quả các lần chạy: 
- Hậu nghiệm thường ra số lần lặp nhỏ hơn so với tiên nghiệm khi nhập cùng sai số. 
- Chữ số có nghĩa và chữ số đáng tin khác nhau như thế nào?\
Chữ số có nghĩa không mang theo thông tin về sai số, chứ số đáng tin có mang theo thông tin về sai số trong đấy. 
Chữ số có nghĩa là chữ số khác 0 kể từ bên trái. 
- 

In [2]:
a = 1e-10

In [3]:
a

1e-10